runtime/src/qio/qio.c

```diff
@@ -2871,46 +2875,117 @@ qioerr _qio_buffered_write(qio_channel_t* ch, const void* ptr, ssize_t len, ssiz
   int64_t remaining = len;
   qioerr err;
   int eof = 0;
+  qio_method_t method = (qio_method_t) (ch->hints & QIO_METHODMASK);
+  ssize_t num_written;
+  size_t num_written_u;
 
   // Include whatever data we got in cached_cur/cached_end
   //_qio_buffered_advance_cached(ch);
 
-  // handle channel position beyond end.
-  //if( _right_mark_start(ch) >= ch->end_pos ) return QIO_EEOF;
-  if (qio_channel_offset_unlocked(ch) >= ch->end_pos) return QIO_EEOF;
+  // if possible make a direct system call instead of buffering
+  if (
+    len >= qio_write_unbuffered_threshold && // the write is large enough
+    method != QIO_METHOD_MMAP &&             // we aren't using mmap
+    method != QIO_METHOD_MEMORY &&           // we aren't using mem
+    ch->mark_cur == 0 &&                     // not waiting for a commit/revert
+    ch->chan_info == NULL                    // there is no IO plugin
+  ) {
+    // printf("Unbuffered Write -----\n");
+
+    // only write up to EOF
+    if ( ch->end_pos < INT64_MAX && _right_mark_start(ch) + len > ch->end_pos ) {
+      remaining = ch->end_pos - _right_mark_start(ch);
+    }
 
-  while ((remaining > 0) && !eof) {
-    if ((ch->bufIoMax > 0) && (remaining > ch->bufIoMax)) {
-      toWrite = ch->bufIoMax;
-    } else {
-      toWrite = remaining;
+    // flush the write-behind portion of the buffer before attempting to write more
+    err = _qio_channel_flush_qio_unlocked(ch);
+    if( err ) goto error;
+
+    // write with a direct system call (may require multiple iterations for write/pwrite)
+    while ( remaining > 0 ) {
+      // printf("remaining: %lld \n", remaining);
+      fflush(stdout);
```

Comment:
Pretty sure this should be commented out (and, generally speaking, better practice today would be to remove the debug printf calls instead of just commenting them out, but they are present in a lot of the qio code).

Reply:
I plan to remove all of my debugging stuff once I'm sure my changes are correct.